{{ cookiecutter.project_slug }}/Dockerfile

```diff
@@ -0,0 +1,64 @@
+#
+# Base image with defaults for all stages
+FROM registry.access.redhat.com/ubi9/python-{{ cookiecutter.python_version.replace(".", "") }} AS base
+
+# Keep this version tag in sync with pyproject.toml
+LABEL konflux.additional-tags="{{ cookiecutter.version }}"
+COPY LICENSE /licenses/
+
+
+#
+# Builder image
+#
+FROM base AS builder
+COPY --from=ghcr.io/astral-sh/uv:0.4.29@sha256:ebb10c5178c7a357d80527f3371e7038561c26234e8a0bb323ea1f2ce8a694b7 /uv /bin/uv
+
+ENV \
+    # use venv from ubi image
+    UV_PROJECT_ENVIRONMENT="/opt/app-root" \
+    # compile bytecode for faster startup
+    UV_COMPILE_BYTECODE="true" \
+    # disable uv cache. it doesn't make sense in a container
+    UV_NO_CACHE=true
+
+COPY pyproject.toml uv.lock ./
+# Test lock file is up to date
+RUN uv lock --locked
+# Install the project dependencies
+RUN uv sync --frozen --no-install-project --no-group dev
+
+COPY README.md ./
+COPY {{ cookiecutter.project_module }} ./{{ cookiecutter.project_module }}
+RUN uv sync --frozen --no-group dev
+
+
+#
+# Test image
+#
+FROM builder AS test
+
+COPY Makefile ./
+RUN uv sync --frozen
+
+COPY tests ./tests
+RUN make test
+
+
+{%- if cookiecutter.deployment == 'pypi' %}
+#
+# PyPI publish image
+#
+FROM test AS pypi
+# TODO: Add the PyPI credentials from the CI/CD secrets (PYPI_USERNAME, PYPI_PASSWORD)
+RUN make pypi
+{%- endif %}
+
+
+{%- if cookiecutter.deployment == 'openshift' %}
+#
+# Production image
+#
+FROM base AS prod
```

Comment:
Should this stage be built from builder instead? It is my understanding that the base stage has not built the project so running this from base will cause the project to be built and dependencies to be installed every time the image is run?

Reply:
🤦🏻 I should have read the next line twice before thinking!